### PR TITLE
Append Sponge support!

### PR DESCRIPTION
**Pull request by nikosgram**
Translate all commands and tasks for Sponge API.

The only staff where I have ignored is the updated signs and the GUI
inventory. Also some code is duplicated on Bukkit and Bungee, can we
move this code to Commons?

### CODE REVIEW

**Comment by minecrafter** [comment]
This is obviously incomplete. I can't accept this in its present form.

File: sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+        listingUpdateTask = new ListingUpdateTask(this);
+        if (apiClient != null) {
+            getLogger().info("Fetching all server packages...");
+            listingUpdateTask.run();
+
+            Sponge.getScheduler().createTaskBuilder()
+                    .execute(listingUpdateTask)
+                    .delay(12, TimeUnit.SECONDS)
+                    .interval(12, TimeUnit.SECONDS)
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java]: Why 12 seconds? This is supposed to be every 10 minutes.

File: sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                20,
+                TimeUnit.MILLISECONDS
+        );
+
+        commandExecutor = new QueuedCommandExecutor(platform);
+
+        Sponge.getScheduler().createTaskBuilder()
+                .execute((Runnable) commandExecutor)
+                .delay(1, TimeUnit.MILLISECONDS)
+                .interval(1, TimeUnit.MILLISECONDS)
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java]: This is supposed to run every tick

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        executeAsyncLater(
+                runnable,
+                0,
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    @Override
+    public void executeAsyncLater(Runnable runnable, long time, TimeUnit unit) {
+        Sponge.getScheduler().createAsyncExecutor(plugin).schedule(
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Use native Sponge API here

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    public void dispatchCommand(String command) {
+        Sponge.getCommandManager().process(
+                Sponge.getServer().getConsole(),
+                command
+        );
+    }
+
+    @Override
+    public void executeAsync(Runnable runnable) {
+        executeAsyncLater(
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Use native Sponge API here

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Sponge.getScheduler().createAsyncExecutor(plugin).schedule(
+                runnable,
+                time,
+                unit
+        ).run();
+    }
+
+    @Override
+    public void executeBlocking(Runnable runnable) {
+        executeBlockingLater(
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Use native Sponge API here

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                runnable,
+                0,
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    @Override
+    public void executeBlockingLater(Runnable runnable, long time, TimeUnit unit) {
+        Sponge.getScheduler().createSyncExecutor(plugin).schedule(
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Use native Sponge API here

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        ).size();
+    }
+
+    @Override
+    public void log(Level level, String message) {
+        switch (level.getName()) {
+            case "WARNING":
+                plugin.getLogger().warn(message);
+                break;
+            default:
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Doesn't account for SEVERE messages

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    @Override
+    public void log(Level level, String message, Throwable throwable) {
+        switch (level.getName()) {
+            case "WARNING":
+                plugin.getLogger().warn(
+                        message,
+                        throwable
+                );
+                break;
+            default:
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Doesn't account for SEVERE messages

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    @Override
+    public CommandExecutor getExecutor() {
+        return plugin.getCommandExecutor();
+    }
+
+    private Optional<Player> convert(QueuedPlayer player) {
+        return Sponge.getServer().getPlayer(
+                UUID.fromString(
+                        player.getUuid()
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Does not handle offline mode. Look at how Bukkit and BungeeCord handles this

File: sponge/src/main/net/buycraft/plugin/sponge/command/ReportSubCommand.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        writer.println("Operating system: " + os);
+        writer.println("Java version: " + javaVersion);
+        writer.println("Server version: " + serverVersion);
+        writer.println("Server IP and port: " + serverIP + " / " + serverPort);
+        writer.println("Online mode: " + Sponge.getServer().getOnlineMode());
+        writer.println();
+
+        writer.println("### Plugin Information ###");
+        writer.println("Plugin version: " + buycraftVersion);
+        writer.println("Platform: Bukkit");
```
Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/command/ReportSubCommand.java]: This is not Bukkit

**Comment by nikosgram** [comment]
@minecrafter you can create a new branch to put it in there and develop the shortages!

File: sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java (hunk 8 above)
Comment (nikosgram) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: Sponge has not support for SEVERE mode in logger!

Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/SpongeBuycraftPlatform.java]: SLF4J offers `Logger.error()` which is equivalent.

**Comment by minecrafter** [comment]
Some of the code could be shared for better use, but much of this is platform-specific code that is not easily shared.

Generally, if you propose a PR, then you're responsible for its contents, and I found a few rather obvious flaws.

File: sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java (hunk 2 above)
Comment (nikosgram) [inline sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java]: Ticks not supported!

image

Comment (minecrafter) [inline sponge/src/main/net/buycraft/plugin/sponge/BuycraftPlugin.java]: Sponge offers `.delayTicks()` and `.intervalTicks()`.

**Comment by nikosgram** [comment]
I'm trying to fix  this flaws. Soz about that!

**Comment by meyerzinn** [comment]
I already almost finished my port...

**Comment by nikosgram** [comment]
Last test for now!

screenshot 46

**Comment by meyerzinn** [comment]
@nikosgram13 You do realize a port is almost completed, no?

File: sponge/src/main/java/net/buycraft/plugin/sponge/command/ReportSubCommand.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        writer.println("Players in queue: " + plugin.getDuePlayerFetcher().getDuePlayers());
+        writer.println("Listing update last completed: " + plugin.getListingUpdateTask().getLastUpdate());
+        writer.println("Listing: ");
+
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        gson.toJson(plugin.getListingUpdateTask().getListing(), writer);
+
+        Sponge.getScheduler().createAsyncExecutor(plugin).execute(() -> {
```
Comment (minecrafter) [inline sponge/src/main/java/net/buycraft/plugin/sponge/command/ReportSubCommand.java]: Use native Sponge API here

**Comment by nikosgram** [comment]
It's ok @20zinnm, I know!

File: sponge/src/main/java/net/buycraft/plugin/sponge/command/ReportSubCommand.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        sender.sendMessage(Text.builder("Please wait...").color(TextColors.GOLD).build());
+
+        final StringWriter out = new StringWriter();
+        final PrintWriter writer = new PrintWriter(out, true);
+
+        String date = new Date().toString();
+        final String os = System.getProperty("os.name") + " | " + System.getProperty("os.version") + " | " + System.getProperty("os.arch");
+        String javaVersion = System.getProperty("java.version") + " | " + System.getProperty("java.vendor");
+        String serverVersion = Sponge.getPlatform().getImplementation().getName();
+        String serverIP = "";//TODO
```
Comment (minecrafter) [inline sponge/src/main/java/net/buycraft/plugin/sponge/command/ReportSubCommand.java]: Use `Sponge.getServer().getBoundAddress()`

Comment (nikosgram) [inline sponge/src/main/java/net/buycraft/plugin/sponge/command/ReportSubCommand.java]:
``` java
address.getAddress().toString(); //or
address.getHostName();
```

File: sponge/src/main/java/net/buycraft/plugin/sponge/logging/BugsnagNilLogger.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package net.buycraft.plugin.sponge.logging;
+
+import com.bugsnag.Logger;
+
+/**
+ * Created on 3/3/2016 for BuycraftX. Read the copyrights before change the code!
```
Comment (mbax) [inline sponge/src/main/java/net/buycraft/plugin/sponge/logging/BugsnagNilLogger.java]: But why?

Comment (nikosgram) [inline sponge/src/main/java/net/buycraft/plugin/sponge/logging/BugsnagNilLogger.java]: Soz, my IDE create it!

Comment (minecrafter) [inline sponge/src/main/java/net/buycraft/plugin/sponge/logging/BugsnagNilLogger.java]: The logger that BuycraftX uses doesn't have `super.<method>()` statements littered around.

As for the license, it's MIT.

File: sponge/src/main/java/net/buycraft/plugin/sponge/util/KeenUtils.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+public class KeenUtils {
+    public static void postServerInformation(BuycraftPlugin plugin) {
+        Map<String, Object> serverData = new LinkedHashMap<>();
+        Map<String, Object> accountData = new LinkedHashMap<>();
+        Map<String, Object> pluginData = new LinkedHashMap<>();
+
+        // Server data
+        serverData.put("id", plugin.getServerInformation().getServer().getId());
+        serverData.put("platform", "bukkit");
```
Comment (minecrafter) [inline sponge/src/main/java/net/buycraft/plugin/sponge/util/KeenUtils.java]: This isn't Bukkit :)

**Comment by minecrafter** [comment]
I've decided on finishing @20zinnm's port instead, but thanks for your contribution.

**Pull request closed without merging**